snuba/datasets/spans_processor.py

```diff
@@ -0,0 +1,87 @@
+import uuid
+from datetime import datetime
+from typing import Any, List, MutableMapping, Optional, Tuple
+
+from sentry_relay.consts import SPAN_STATUS_NAME_TO_CODE
+
+from snuba.datasets.events_format import (
+    enforce_retention,
+    extract_extra_tags,
+    extract_project_id,
+)
+from snuba.processor import (
+    InsertBatch,
+    MessageProcessor,
+    ProcessedMessage,
+    _as_dict_safe,
+    _ensure_valid_date,
+    _unicodify,
+)
+
+UNKNOWN_SPAN_STATUS = 2
+
+
+class SpansMessageProcessor(MessageProcessor):
+    def __extract_timestamp(self, field: float) -> Tuple[datetime, int]:
+        timestamp = _ensure_valid_date(datetime.fromtimestamp(field))
+        if timestamp is None:
+            timestamp = datetime.utcnow()
+        nanoseconds = int(timestamp.microsecond * 1000)
+        return (timestamp, nanoseconds)
+
+    def process_message(self, message, metadata) -> Optional[ProcessedMessage]:
+        if not (isinstance(message, (list, tuple)) and len(message) >= 2):
+            return None
+        version = message[0]
+        if version not in (0, 1, 2):
+            return None
+        type_, event = message[1:3]
+        if type_ != "insert":
+            return None
+
+        data = event["data"]
+        event_type = data.get("type")
+        if event_type != "transaction":
+            return None
+
+        ret: List[MutableMapping[str, Any]] = []
+        spans = data.get("spans", [])
+        for span in spans:
+            processed: MutableMapping[str, Any] = {"deleted": 0}
+            extract_project_id(processed, event)
+            processed["retention_days"] = enforce_retention(
+                event, datetime.fromtimestamp(data["timestamp"]),
+            )
+            processed["transaction_id"] = str(uuid.UUID(event["event_id"]))
+            processed["span_id"] = int(span["span_id"], 16)
+            processed["transaction_name"] = _unicodify(data.get("transaction") or "")
+            processed["parent_span_id"] = int(span["parent_span_id"], 16)
```

Comment:
I think parent_span_id can be empty for the root span.

Reply:
How so? Then what is the span id of the root span ? Also span_id on the transaction table is not nullable.

Reply:
Per our discussion: Every transaction is also a span, and so has a span_id. But a transaction doesn't necessarily have a parent_span_id.

Reply:
Kept it not nullable. We will store 0 for the root spans. It should give us better performance during queries

Reply:
Actually turned it to nullable